File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/core/Picker.kt
// Picking the next card, and the progress through today's lists that it keeps.
package com.wildduck.azkar.core

import kotlinx.datetime.LocalDateTime

enum class Session(val raw: String) { Sabah("sabah"), Masaa("masaa"), General("general") }

data class Card(
    val zikr: Zikr,
    val session: Session,
    /** 1-based position within the morning/evening list; null for general azkar. */
    val position: Int? = null,
    val total: Int? = null,
)

/** Kept between launches so the morning/evening lists continue where they left off today. */
data class AppState(
    val day: String = "",
    val sabah: Int = 0,
    val masaa: Int = 0,
    val general: Int = 0,
    val lastGeneral: Int? = null,
    val paused: Boolean = false,
)

/** The picked card (null when there is nothing to show) and the state after picking. */
data class Pick(val card: Card?, val state: AppState)

object Picker {
    /**
     * Inside the morning (evening) window, walks that list in order once per day; otherwise, or once the list is
     * done, picks from `general` (one tap each unless `repeatGeneral`). `random(n)` must return 0 until n.
     */
    fun next(at: LocalDateTime, config: Config, library: Library, state: AppState, random: (Int) -> Int): Pick {
        val today = dayKey(at)
        var st = if (state.day == today) state else state.copy(day = today, sabah = 0, masaa = 0)
        val minute = minuteOfDay(at)

        val sabah = library.sabah
        if (config.sabah?.contains(minute) == true && st.sabah < sabah.size) {
            val i = st.sabah
            return Pick(Card(sabah[i], Session.Sabah, i + 1, sabah.size), st.copy(sabah = i + 1))
        }
        val masaa = library.masaa
        if (config.masaa?.contains(minute) == true && st.masaa < masaa.size) {
            val i = st.masaa
            return Pick(Card(masaa[i], Session.Masaa, i + 1, masaa.size), st.copy(masaa = i + 1))
        }

        val general = library.general
        if (general.isEmpty()) return Pick(null, st)
        val i: Int
        when (config.order) {
            Order.Sequential -> {
                i = st.general % general.size
                st = st.copy(general = i + 1)
            }
            Order.Random -> {
                val last = st.lastGeneral
                i = when {
                    general.size == 1 -> 0
                    // Skip `last` so the same zikr never shows twice in a row.
                    last != null && last < general.size -> random(general.size - 1).let { r -> if (r >= last) r + 1 else r }
                    else -> random(general.size)
                }
            }
        }
        val zikr = if (config.repeatGeneral) general[i] else general[i].copy(count = 1)
        return Pick(Card(zikr, Session.General), st.copy(lastGeneral = i))
    }
}

File: azkar/apple/Core/Picker.swift
// Picking the next card, and the progress through today's lists that it keeps.
import Foundation

enum Session: String, Codable { case sabah, masaa, general }

struct Card: Equatable {
  var zikr: Zikr
  var session: Session
  /// 1-based position within the morning/evening list; nil for general azkar.
  var position: Int?
  var total: Int?
}

/// Persisted between launches so the morning/evening lists continue where they left off today.
struct AppState: Codable, Equatable {
  var day = ""
  var sabah = 0
  var masaa = 0
  var general = 0
  var lastGeneral: Int?
  var paused = false
}

enum Picker {
  /// Inside the morning (evening) window, walks that list in order once per day; otherwise,
  /// or once the list is done, picks from `general` (one tap each unless `repeatGeneral`).
  /// `random(n)` must return 0..<n.
  static func next(
    at date: Date, calendar: Calendar, config: Config, library: Library,
    state: inout AppState, random: (Int) -> Int
  ) -> Card? {
    let today = dayKey(date, calendar)
    if state.day != today {
      state.day = today
      state.sabah = 0
      state.masaa = 0
    }

    let minute = minuteOfDay(date, calendar)
    let timed: [(Session, TimeWindow?, [Zikr], WritableKeyPath<AppState, Int>)] = [
      (.sabah, config.sabah, library.sabah, \.sabah),
      (.masaa, config.masaa, library.masaa, \.masaa),
    ]
    for (session, window, list, progress) in timed {
      guard let window, window.contains(minuteOfDay: minute), state[keyPath: progress] < list.count else { continue }
      let i = state[keyPath: progress]
      state[keyPath: progress] += 1
      return Card(zikr: list[i], session: session, position: i + 1, total: list.count)
    }

    let general = library.general
    guard !general.isEmpty else { return nil }
    let i: Int
    switch config.order {
    case .sequential:
      i = state.general % general.count
      state.general = i + 1
    case .random:
      if general.count == 1 {
        i = 0
      } else if let last = state.lastGeneral, last < general.count {
        let r = random(general.count - 1)  // skip `last` so the same zikr never shows twice in a row
        i = r >= last ? r + 1 : r
      } else {
        i = random(general.count)
      }
    }
    state.lastGeneral = i
    var zikr = general[i]
    if !config.repeatGeneral { zikr.count = 1 }
    return Card(zikr: zikr, session: .general, position: nil, total: nil)
  }
}
